cprover_bindings/src/irep/goto_binary_serde.rs

```diff
@@ -0,0 +1,1365 @@
+        let named_sub: Vec<(usize, usize)> = irep
+            .named_sub
+            .iter()
+            .map(|(key, value)| (self.number_irep_id(key).number, self.number_irep(value).number))
+            .collect();
+        let key = IrepKey::new(id, &sub, &named_sub);
+        self.get_or_insert(&key)
+    }
+
+    /// Gets the existing [NumberedIrep] from the [IrepKey] or inserts a fresh
+    /// one and returns it.
+    fn get_or_insert(&mut self, key: &IrepKey) -> NumberedIrep {
+        if let Some(number) = self.cache.get(key) {
+            // Return the NumberedIrep from the inverse cache
+            return self.inv_cache.index[*number];
+        }
+        // This is where the key gets its unique number assigneds.
+        let number = self.inv_cache.add_key(&key);
+        self.cache.insert(key.clone(), number);
+        self.inv_cache.index[number]
+    }
+
+    /// Returns the unique number of the `id` field of the given [NumberedIrep].
+    fn id(&self, numbered_irep: &NumberedIrep) -> NumberedString {
+        self.inv_string_cache[self.inv_cache.keys[numbered_irep.start_index]]
+    }
+
+    /// Returns `#sub`, the number of `sub` [Irep]s of the given [NumberedIrep].
+    /// It is found at `numbered_irep.start_index + 1` in the inverse cache.
+    fn nof_sub(&self, numbered_irep: &NumberedIrep) -> usize {
+        self.inv_cache.keys[numbered_irep.start_index + 1]
+    }
+
+    /// Returns the [NumberedIrep] for the ith `sub` of the given [NumberedIrep].
+    fn sub(&self, numbered_irep: &NumberedIrep, sub_idx: usize) -> NumberedIrep {
+        let sub_number = self.inv_cache.keys[numbered_irep.start_index + sub_idx + 2];
+        self.inv_cache.index[sub_number]
+    }
+
+    /// Returns `#named_sub`, the number of named subs of the given [NumberedIrep].
+    /// It is found at `numbered_irep.start_index + #sub + 2` in the inverse cache.
+    fn nof_named_sub(&self, numbered_irep: &NumberedIrep) -> usize {
+        self.inv_cache.keys[numbered_irep.start_index + self.nof_sub(numbered_irep) + 2]
+    }
+
+    /// Returns the pair of [NumberedString] and [NumberedIrep] for the named
+    /// sub number `i` of this [NumberedIrep].
+    fn named_sub(
+        &self,
+        numbered_irep: &NumberedIrep,
+        named_sub_idx: usize,
+    ) -> (NumberedString, NumberedIrep) {
+        let start_index =
+            numbered_irep.start_index + self.nof_sub(numbered_irep) + 2 * named_sub_idx + 3;
+        (
+            self.inv_string_cache[self.inv_cache.keys[start_index]],
+            self.inv_cache.index[self.inv_cache.keys[start_index + 1]],
+        )
+    }
+}
+
+/// A uniquely numbered [Irep].
+/// A NumberedIrep can be viewed as a generational index into a [IrepNumbering] instance.
+#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
+struct NumberedIrep {
+    /// The unique number of this NumberedIrep.
+    number: usize,
```

Comment:
I'll look into this, but I think that this would prevent using them as indices into the inverse caches. The inverse cache stores a mix of irep numbers, string numbers, vector and map lengths in a flat usize vector.